pages/k8s/snaps.md

```diff
@@ -0,0 +1,189 @@
+context:
+  title: "Kubernetes Snaps"
+  description: How to manage and refresh snaps
+keywords: snaps
+tags: [reference]
+sidebar: k8smain-sidebar
+permalink: snaps.html
+layout: [base, ubuntu-com]
+toc: False
+---
+
+**Charmed Kubernetes** makes extensive use of [snaps], both for Kubernetes
+services and for client-side tools to operate a cluster. This page details some
+of the Kubernetes-related snaps available and how they are used.
+
+## Client-side Snaps
+
+The following client tools are available. Note that it is not _required_ that
+you use the Snap version of these tools to work with Charmed Kubernetes, they
+are just provided as a maintained and tested way of keeping up to date.
+
+`kubectl` - client for running commands on a Kubernetes cluster. [Docs][kubectl] &#124; [Snap][kubectl-snap]
+
+`kubeadm` - a tool for bootstrapping a Kubernetes cluster. [Docs][kubeadm] &#124; [Snap][kubeadm-snap]
+
+`kubefed` - for running commands on federated clusters. [Docs][kubefed] &#124; [Snap][kubefed-snap]
+
+These are all installed in the same way. For example:
+
+```bash
+sudo snap install kubectl --classic
+```
+
+## Server Snaps
+
+When installing **Charmed Kubernetes**, various necessary services are actually
+installed from snap packages. The installation and management of these snaps is
+undertaken by the charms; there is no need for a user to interact with the
+snaps directly. The details here are provided for information only.
+
+
+| Snap | Type | Store page |
+| kube-apiserver | strict | <https://snapcraft.io/kube-apiserver> |
+| kube-controller-manager | strict  | <https://snapcraft.io/kube-controller-manager>  |
+| kube-proxy | classic | <https://snapcraft.io/kube-proxy> |
+| kube-scheduler  |  strict  | <https://snapcraft.io/kube-scheduler>  |
+| kubelet  | classic  | <https://snapcraft.io/kubelet>  |
+
+### Example: kube-apiserver
+
+We will use kube-apiserver as an example. The other services generally work the
+same way.
+
+Install with `snap install`:
+
+```
+sudo snap install kube-apiserver
+```
+
+This creates a systemd service named `snap.kube-apiserver.daemon`. Initially,
+it will be in an error state because it's missing important configuration:
+
+Running:
+
+```bash
+systemctl status snap.kube-apiserver.daemon
+```
+
+... will return status similar to:
+
+```no-highlight
+● snap.kube-apiserver.daemon.service - Service for snap application kube-apiserver.daemon
+   Loaded: loaded (/etc/systemd/system/snap.kube-apiserver.daemon.service; enabled; vendor preset: enabled)
+   Active: inactive (dead) (Result: exit-code) since Fri 2020-09-01 15:54:39 UTC; 11s ago
+   ...
+```
+
+The snap can be configured using `snap set`:
+
+```bash
+sudo snap set kube-apiserver \
+  etcd-servers=https://172.31.9.254:2379 \
+  etcd-certfile=/root/certs/client.crt \
+  etcd-keyfile=/root/certs/client.key \
+  etcd-cafile=/root/certs/ca.crt \
+  service-cluster-ip-range=10.123.123.0/24 \
+  cert-dir=/root/certs
```

Comment:
This is the old style of arg-setting that we would like to deprecate. Nowadays we do something like:

```
sudo snap set kube-apiserver args="
--etcd-servers=https://172.31.9.254:2379
--etcd-certfile=/root/certs/client.crt
--etcd-keyfile=/root/certs/client.key
--etcd-cafile=/root/certs/ca.crt
--service-cluster-ip-range=10.123.123.0/24
--cert-dir=/root/certs
"
```
^ I haven't checked this is 100% correct but I can check later.

Reply:
Checked this, it works.